bazel/external/icuuc.BUILD

```diff
@@ -0,0 +1,59 @@
+load("@rules_cc//cc:defs.bzl", "cc_library")
+
+licenses(["notice"])  # Apache 2
+
+exports_files([
+    "icu4c/LICENSE",
+    "icu4j/main/shared/licenses/LICENSE",
+])
+
+icuuc_copts = [
+    "-DU_STATIC_IMPLEMENTATION",
+    "-DU_COMMON_IMPLEMENTATION",
+    "-DU_HAVE_STD_ATOMICS",
+] + select({
+    "@envoy//bazel:apple": [
+        "-Wno-shorten-64-to-32",
+        "-Wno-unused-variable",
+    ],
+    "@envoy//bazel:windows_x86_64": [
+        "/utf-8",
+        "/DLOCALE_ALLOW_NEUTRAL_NAMES=0",
+    ],
+    # TODO(dio): Add "@envoy//bazel:android" when we have it.
+    # "@envoy//bazel:android": [
+    #     "-fdata-sections",
+    #     "-DU_HAVE_NL_LANGINFO_CODESET=0",
+    #     "-Wno-deprecated-declarations",
+    # ],
+    "//conditions:default": [],
+})
+
+cc_library(
+    name = "headers",
+    hdrs = glob(["icu4c/source/common/unicode/*.h"]),
+    includes = ["icu4c/source/common"],
+    visibility = ["//visibility:public"],
+)
+
+cc_library(
+    name = "common",
+    hdrs = glob(["icu4c/source/common/unicode/*.h"]),
+    includes = ["icu4c/source/common"],
+    visibility = ["//visibility:public"],
+    deps = [":icuuc"],
+)
+
+cc_library(
+    name = "icuuc",
+    srcs = glob([
+        "icu4c/source/common/*.c",
+        "icu4c/source/common/*.cpp",
+        "icu4c/source/stubdata/*.cpp",
+    ]),
+    hdrs = glob(["icu4c/source/common/*.h"]),
+    copts = icuuc_copts,
+    tags = ["requires-rtti"],
+    visibility = ["//visibility:private"],
+    deps = [":headers"],
+)
```

Comment:
@htuch: This is an autoconf dependency, can we just use `foreign_cc` to avoid having to hand-maintain a bespoke BUILD file for the project?
@dio: I'm testing it, but it seems the build time is significantly increasing.
@htuch: Interesting, I wonder if it's doing something like building tests or additional libraries. Might be possible to control with some autoconf flags...
@dio: I believe I have turned off everything (that makes sense, like examples, tests, even tools). And it is tricky to force using msvc (when using configure it tends to use mingw and it is failed at configure stage, and I don't think we want to mix mingw and msvc).

Ref: https://github.com/envoyproxy/envoy/pull/10599#discussion_r403338250

Reply:
I'm confused, what is an autoconf dependency? If one is introduced by this patch, that will break the contribution of all four engineers working on the Windows port by breaking master. The one I am familiar with is the opentracing dependency which we have excluded for now, but that only breaks tracing extensions. Breaking http/1 seems unwise. Otherwise the basic idea of this bazel BUILD hack looks entirely reasonable, and there is no expectation it would be fixed upstream, because we are cherry-picking a project for subcomponents that don't have any independent project management and don't appear to be supported stand-alone.

Reply:
I'm OK with this BUILD (as it isn't a huge patch). In terms of being an autoconf dependency, I'm basically saying it should build under `rules_foreign_cc`, which knows how to build autoconf. 

`gperftools` is an autoconf dep today that we build this way, see https://github.com/envoyproxy/envoy/blob/master/bazel/foreign_cc/BUILD#L13.

Are you saying in the future, we can't take any external dependency that is autoconf based?

Reply:
Correct, autoconf is gnu tooling and while it runs in tandem the pseudo-unix cygwin toolchain there is no Windows port, and it's likely to be a thorn in the side of other odd SDKs on mobile and such.

Most every project building on autoconf offers an alternative mechanism for some windows build, however well thought out and supported, or not. external/org_unicode_icuuc/icu4c/source/allinone/ is the Windows build logic by the icu4c maintainers.

Most projects are moving in the direction of supporting CMake for a much simpler cross-platform experience, even as they retain autoconf. This makes the build agnostic of the flavor of Visual Studio and even msvc vs llvm clang-cl.

For Envoy, maintaining a bazel BUILD file for this project makes perfect sense, since the lib objects all match the compiler and link choices of this project. We need to evaluate what the autoconf results produce in terms of source file toggles and changes, even if we avoid these long-term.

For upstream, propose a CMakeFiles.txt to cover their existing autoconf determinations. Our rules_foreign_cc support in bazel is much better with CMake than less-predictable autoconf output. This probably results in the most flexibility to the maintainers and would be our simplest solution short of this short-term BUILD maintenance.

Reply:
@wrowe can you propose some amendment to https://github.com/envoyproxy/envoy/blob/master/bazel/EXTERNAL_DEPS.md based on this?

I don't think it sounds unreasonable, but we should take this through a review process and ensure folks are aligned, as it's a major project restriction.

How come gperftools isn't impacting you? Don't you use tcmalloc?

Reply:
tcmalloc is disabled on Windows if you look at .bazelrc

Reply:
gperftools as of 2015 was being ported to Windows using the tcmalloc_minimal implementation. We have have documented this exclusion for Windows both in .bazelrc and in bazel/foreign_cc/BUILD.

Until bazel test //test/... is largely building clean with all currently tagged "fails_on_windows", gperftools hasn't been and won't be a particularly high priority. The same is true for opentracing and all tracing extensions on Windows. 

THE major difference between all that and this proposed enhancement is that the envoyproxy is really nothing without a url parser.

Reply:
This is fine, but I would like to get this down in writing, so that we can ensure everyone is on-board and don't relitigate this each time we add a new dependency. Can you do the PR to https://github.com/envoyproxy/envoy/blob/master/bazel/EXTERNAL_DEPS.md  arguing to remove autoconf in rules_foreign_cc for Windows reasons for all deps that are core across platforms?